src/react/reconcilation.js

```diff
@@ -107,29 +149,59 @@ export class Reconciler {
   ) {
     let value;
     let childContext = context;
+
+    // first we check if it's a class component
     if (valueIsClassComponent(this.realm, componentType)) {
-      if (branchStatus !== "ROOT") {
-        throw new ExpectedBailOut("only class components at the root of __registerReactComponentRoot() are supported");
+      let classComponentTypes = this.realm.react.classComponentTypes;
+      // We first need to know what type of class component we're dealing with.
+      // A "simple" class component is defined as:
+      //
+      // - having only a "render" method or many method, i.e. render(), _renderHeader(), _renderFooter()
+      // - having no lifecycle events
+      // - having no state
+      // - having no instance variables
+      //
+      // the only things a class component should be able to access on "this" are:
+      // - this.props
+      // - this.context
+      // - this._someRenderMethodX() etc
+      //
+      // Otherwise, the class component is a "complex" one.
+      // To begin with, we don't know what type of component it is, so we try and render it as if it were
+      // a simple component using the above heuristics. If an error occurs during this process, we assume
+      // that the class wasn't simple, then try again with the "complex" heuristics.
+      if (!classComponentTypes.has(componentType)) {
+        try {
+          value = this._renderSimpleClassComponent(componentType, props, context, branchStatus, branchState);
+          classComponentTypes.set(componentType, "SIMPLE");
+        } catch (error) {
+          // if we get back a SimpleClassBailOut error, we know that this class component
+          // wasn't a simple one and is likely to be a complex class component instead
+          if (error instanceof SimpleClassBailOut) {
+            // mark the component as not simple and continue
+            classComponentTypes.set(componentType, "COMPLEX");
+          } else {
+            // else we rethrow the error
+            throw error;
+          }
+        }
+      } else if (classComponentTypes.get(componentType) === "SIMPLE") {
+        value = this._renderSimpleClassComponent(componentType, props, context, branchStatus, branchState);
```

Comment:
This still has the same problem that if we have two components which will pass in one case and not pass in another case, then it will seem pretty random when it does and doesn't switch to the complex mode.

```js
class SimpleOrComplex extends Component {
  render() {
    if (this.props.data) {
      this._cache = computeString(this.props.data, this._cache);
      return <div>{this._cache}</div>;
    }
    return <div>Default</div>;
  }
}
```

```js
function Foo() {
  return <SimpleOrComplex data="hello" />;
}
function Bar() {
  return <SimpleOrComplex />;
}
function App({c}) {
  return c ? <Foo /> : <Bar />;
}
```

Depending on if we happen to visit Foo or Bar first, or if the user inverts the condition in App, this will either fail or not.

I'd rather have this be more predictable. E.g. always try the simple mode first and just remove the `classComponentTypes` cache.

Reply:
Got ya.